src/gadgets/CensorshipResistanceGadget.sol

```diff
@@ -0,0 +1,84 @@
+pragma solidity ^0.8.19;
+
+import {AttestationGadget} from "src/gadgets/AttestationGadget.sol";
+
+import {IAndromeda} from "src/IAndromeda.sol";
+
+abstract contract CensorshipResistanceGadget is AttestationGadget {
+    function Suave() internal virtual returns (IAndromeda);
+
+    // require explicit onchain challenge on kettle restart
+    mapping(bytes32 => bool) private cr_challenges;
+
+    function offchain_restart_challenge() public returns (bytes32 challenge, bytes memory attestation) {
+        require(!cr_challenges[challenge]); /* already challenged */
+        challenge = _get_local_challenge(); /* do not override previous local challenge to prevent dos */
+        if (challenge == bytes32(0)) {
+            challenge = Suave().localRandom();
+            _set_local_challenge(challenge);
+        }
+        attestation = offchain_attest(this.onchain_restart_challenge.selector, abi.encodePacked(challenge));
+    }
+
+    function onchain_restart_challenge(bytes32 challenge, bytes memory attestation)
+        public
+        onchain_verify(abi.encodePacked(challenge), attestation)
+    {
+        cr_challenges[challenge] = true;
+    }
```

Comment:
I know that this check was done in the offchain_restart_challenge, but shouldn't it be here too?
`require(!cr_challenges[challenge]);`

Reply:
The call is idempotent so it's fine like this too, matter of preference